First version of Population module

allbackoff opened the pull request:
Implemented methods for Population module, need revising later

allbackoff committed:
First version of Population module

fesiib commented on src/population.py:
```diff
@@ -0,0 +1,58 @@
+import random
+
+class Population():
+    def __init__(self, solutions, fitness):
```
`def __init__(self, solutions = None, fitness = None):`

Since when we create an instance of Population class we may not have specific solutions or fitness

fesiib replied:
`def __init__(self, solutions = None, fitness = None):`

Since when we create an instance of Population class we may not have specific solutions or fitness

fesiib commented on src/population.py:
```diff
@@ -0,0 +1,58 @@
+        self.solutions = solutions
+        self.fitness = fitness
+
+    def initial_population(self, population_size, test_size):
```
Please add a default value for `test_size = 1`, for now. 

We need a more sophisticated way of marking test types (not only test_size). But for now, it is okay...

fesiib commented on src/population.py:
```diff
@@ -0,0 +1,58 @@
+    def initial_population(self, population_size, test_size):
+        population = list()
+        for i in range(population_size):
+            test = list()
```
There is a specific TestCase class for the test, so probably you should create an instance of that class.

fesiib commented on src/population.py:
```diff
@@ -0,0 +1,58 @@
+            for j in range(test_size):
+                test.append(random.randint(-1000, 1000))
+            population.append(test)
+        return population
```
What the hell??

why do you need `self.solutions`?

You have to store all the TestCases in `self.solutions`

fesiib commented on src/population.py:
```diff
@@ -0,0 +1,58 @@
+            population.append(test)
+        return population
+
+    def mutate_1(self, lower, upper, prob):
```
Please look at Class Diagram, it is updated now.

There you will see a new `mutate(self, solution, mode)` function. It is basically the same, i.e just wrapper around these three `mutate_1`, `mutate_2`, `mutate_3`.

Also, please try to follow function parameters and return values from the Class Diagram.

fesiib commented on src/population.py:
```diff
@@ -0,0 +1,58 @@
+        prob -= rate
+        return mutated
+
+    def crossover(self, parent_1, parent_2, rate):
```
Ba pochemu ty zabil na to chto napisano v Class Diagram????

fesiib replied:
Also you do not `rate` in here, it is already handled in `generator.py`

Rainhunter13 replied:
ble oru

fesiib reviewed (changes requested):
Please implement other functions too.

allbackoff committed:
Revised every function based on new Class Diagram

allbackoff commented:
One detail is not resolved in the latest commit: in `tournament_selection()` function, I need the `TestCase.get_fitness()` function and I'm not sure what constraint I should include there as a parameter.

fesiib commented on src/population.py:
```diff
@@ -1,58 +1,56 @@
+    def initial_population(self, population_size: int, test_type: int):
+        for _ in range(population_size):
+            test = TestCase(test_type=test_type)
+            self.solutions.append(test.get_input())
```
`self.solutions.append(test);`

you should append an instance of TestCase, not the input list

fesiib commented on src/population.py:
```diff
@@ -1,58 +1,56 @@
-            if self.fitness[x] > self.fitness[y]:
-                parent = x
+            candidate_1, candidate_2 = random.choices(self.solutions)
+            if candidate_1.get_fitness() > candidate_2.get_fitness():
```
I am not sure rn, but lower the fitness() -> better. bc constraint will be = 0 when it is satisfied. But let's see

fesiib reviewed (changes requested):
Overall good, just fix several mistakes

fesiib commented:
> One detail is not resolved in the latest commit: in `tournament_selection()` function, I need the `TestCase.get_fitness()` function and I'm not sure what constraint I should include there as a parameter.

I forgot to add `constraint` argument for `tournament_selection()`. My bad, you can assume that some instance of `Constraint` is given as argument

allbackoff committed:
1. initial_population: append test, not the input list
2. tournament_selection:lower fitness is better in terms of selection
3. tournament_selection: added constraint as parameter
